api/pkg/transformer/types/encoder/cyclical_encoder.go

```diff
@@ -0,0 +1,312 @@
+		case spec.PeriodType_DAY:
+			period = spec.PeriodType_UNDEFINED
+			max = DayInSec
+		case spec.PeriodType_WEEK:
+			period = spec.PeriodType_UNDEFINED
+			max = WeekInSec
+		case spec.PeriodType_MONTH, spec.PeriodType_QUARTER, spec.PeriodType_HALF, spec.PeriodType_YEAR:
+			period = byEpochTime.Period
+			max = 0
+		default:
+			return nil, fmt.Errorf("invalid or unspported cycle period")
+		}
+
+		return &CyclicalEncoder{
+			Period: period,
+			Min:    min,
+			Max:    max,
+		}, nil
+	}
+
+	return nil, fmt.Errorf("cyclical encoding config invalid or undefined")
+}
+
+func (oe *CyclicalEncoder) Encode(values []interface{}, column string) (map[string]interface{}, error) {
+	encodedCos := make([]interface{}, 0, len(values))
+	encodedSin := make([]interface{}, 0, len(values))
+
+	// config with fixed range
+	if oe.Period == spec.PeriodType_UNDEFINED {
+		period := oe.Max - oe.Min
+		unitAngle := completeAngle / period
+
+		for _, val := range values {
+			//Check if value is missing
+			if val == nil {
+				return nil, fmt.Errorf("missing value")
+			}
+
+			//Check if value is valid
+			valFloat, err := converter.ToFloat64(val)
+			if err != nil {
+				return nil, err
+			}
+
+			//Encode to sin and cos
+			phase := (valFloat - oe.Min) * unitAngle
+			encodedCos = append(encodedCos, math.Cos(phase))
+			encodedSin = append(encodedSin, math.Sin(phase))
+		}
+	} else {
+		//config with variable range, by epoch time (e.g. different days in each month, leap year etc.)
+		for _, val := range values {
+			//Check if value is missing
+			if val == nil {
+				return nil, fmt.Errorf("missing value")
+			}
+
+			//Check if value is valid
+			valInt, err := converter.ToInt64(val)
+			if err != nil {
+				return nil, err
+			}
+
+			//convert epoch time to golang datetime
+			t := time.Unix(valInt, 0).In(time.UTC)
+			shareOfPeriod, err := getCycleTime(oe.Period, t)
+			if err != nil {
+				return nil, err
+			}
+			unitAngle, err := getUnitAngle(oe.Period, t)
+			if err != nil {
+				return nil, err
+			}
+
+			//Encode to sin and cos
+			phase := shareOfPeriod * unitAngle
+			encodedCos = append(encodedCos, math.Cos(phase))
+			encodedSin = append(encodedSin, math.Sin(phase))
+		}
+	}
+
+	return map[string]interface{}{
+		column + "_x": encodedCos,
+		column + "_y": encodedSin,
+	}, nil
+}
+
+//Computes the number of seconds past the beginning of a pre-defined cycle
+//Only works with PeriodType with variable cycle time such as Month, Year etc
+//For period type with fixed cycle time, it is handled differently by encoder and
+//does not need the cycle time to be computed
+func getCycleTime(periodType interface{}, t time.Time) (float64, error) {
+	_, pType := periodType.(spec.PeriodType)
+	if !pType {
+		return 0, fmt.Errorf("invalid type for periodType")
+	}
+
+	switch periodType {
+	case spec.PeriodType_MONTH:
+		dayElapsed := t.Day() - 1
+		hr, min, sec := t.Clock()
+		elapsed := float64(dayElapsed*DayInSec + hr*HourInSec + min*MinInSec + sec)
```

Comment:
can you make new function for `elapsed` ? it looks like it share the same logic with the one below. Maybe `elapsed(time time.Time, dayElapsed int64, periodElapsed int64)` ?